File: src/main/scala/com/vc/medium/L94.scala
package com.vc.medium


class TreeNode94(var _value: Int) {
  var value: Int = _value
  var left: TreeNode94 = null
  var right: TreeNode94 = null
}

object Solution {
  def inorderTraversal(root: TreeNode94): List[Int] = {
    import scala.collection.mutable
    val list = new mutable.ListBuffer[Int]()
    val st = new mutable.Stack[TreeNode94]()
    var current = root
    while(st.nonEmpty || current != null) {
      if(current != null) {
        st.push(current)
        current = current.left
      }
      else {
        current = st.pop
        list += current.value
        current = current.right
      }
    }
    list.toList
  }
}

File: src/main/java/com/vc/medium/L94.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.*;

class L94 {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        TreeNode current = root;
        Stack<TreeNode> st = new Stack<>();
        while(current != null || !st.isEmpty()) {
            if(current != null) {
                st.push(current);
                current = current.left;
            }
            else {
                current = st.pop();
                list.add(current.val);
                current = current.right;
            }
        }
        return list;
    }
}
